File: iOS_JavaScript/iOS-callJavaScript/OC&JavaScript/OCWatchJavaScript/ViewController.m
//
//  ViewController.m
//  OCWatchJavaScript
//
//  Created by 孟跃平 on 2017/6/15.
//  Copyright © 2017年 www.mengyueping.com. All rights reserved.
//

/*
 *  OC利用UIWebView监听JavaScript函数的触发
 */

#import "ViewController.h"

@interface ViewController ()<UIWebViewDelegate,UINavigationControllerDelegate, UIImagePickerControllerDelegate>
@end
@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIWebView *webView = [[UIWebView alloc] initWithFrame:self.view.bounds];
    webView.delegate = self;
    webView.backgroundColor = [UIColor grayColor];
    webView.scalesPageToFit = YES;
    [self.view addSubview:webView];
    
    NSURL *htmlUrl = [[NSBundle mainBundle] URLForResource:@"index" withExtension:@"html"];
    [webView loadRequest:[NSURLRequest requestWithURL:htmlUrl]];
}

//代理 UIWebViewDelegate
- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType
{
    NSString *url = request.URL.absoluteString;
    NSRange range = [url rangeOfString:@"mengyueping.com://"]; //自定义协议
    NSUInteger location = range.location;
    
    if (location != NSNotFound) {
        NSString *str = [url substringFromIndex:(location + range.length)];
        SEL selector = NSSelectorFromString(str);
        //警告：PerformSelector may cause a leak because its selector is unknown
        //http://www.tuicool.com/articles/iu6zuu
        //[self performSelector:selector];
        [self performSelector:selector withObject:nil afterDelay:0.0];
    }
    
    return YES;
}

// iOS原生方法访问相册
- (void)openCamera
{
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    [self presentViewController:picker animated:YES completion:nil];
    
}



@end

File: iOS_JavaScript/iOS-callJavaScript/Swift&JavaScript/SwfitWatchJavaScript/ViewController.swift
//
//  ViewController.swift
//  SwfitWatchJavaScript
//
//  Created by 孟跃平 on 2017/6/15.
//  Copyright © 2017年 www.mengyueping.com. All rights reserved.
//

/*
 *  Swift利用UIWebView监听JavaScript函数的触发
 */

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let webView = UIWebView(frame: view.bounds)
        webView.delegate = self
        webView.backgroundColor = .gray
        webView.scalesPageToFit = true
        view.addSubview(webView)
        
        let url = Bundle.main.url(forResource:"index", withExtension:"html")!
        webView.loadRequest(URLRequest(url: url))
    }
    
    // iOS原生方法访问相册
    @objc fileprivate func openCamera() {
        let pickerVC = UIImagePickerController()
        pickerVC.sourceType = .photoLibrary
        self.present(pickerVC, animated: true, completion: nil)
    }
}

extension ViewController: UIWebViewDelegate {
    
    func webView(_ webView: UIWebView, shouldStartLoadWith request: URLRequest, navigationType: UIWebViewNavigationType) -> Bool {
        
        let url = request.url?.absoluteString
        let range = url?.range(of: "mengyueping.com://") //自定义协议
        guard let _ = range else {
            return true
        }
        
        let str = url?.substring(from: "mengyueping.com://".endIndex)
        guard let selStr = str else {
            return true
        }
        let selector = Selector(selStr)
        perform(selector)
        
        return true
    }
}


